File: galite-core/src/main/java/org/kopi/vkopi/lib/ui/swing/visual/DActor.java
/*
 * Copyright (c) 1990-2021 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id: DActor.java 34961 2016-11-04 17:20:49Z hacheni $
 */

package org.kopi.vkopi.lib.ui.swing.visual;

import java.awt.event.ActionEvent;
import java.awt.event.KeyEvent;

import javax.swing.AbstractAction;
import javax.swing.Action;
import javax.swing.Icon;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.KeyStroke;

import org.kopi.vkopi.lib.ui.swing.base.JMenuButton;
import org.kopi.galite.visual.visual.UActor;
import org.kopi.galite.visual.visual.VActor;

@SuppressWarnings("serial")
public class DActor implements UActor {

  // --------------------------------------------------------------------
  // CONSTRUCTOR
  // --------------------------------------------------------------------

  public DActor(VActor model) {
    this.model = model;
    init();
  }

  // --------------------------------------------------------------------
  // ACCESSORS
  // --------------------------------------------------------------------

  
  public boolean isEnabled() {
    return action.isEnabled();
  }

  
  public void setEnabled(boolean enabled) {
    action.setEnabled(enabled);
  }

  
  public boolean isVisible() {
    return button != null && button.isVisible();
  }

  
  public void setVisible(boolean visible) {
    if (button != null) {
      button.setVisible(visible);
    }
  }

  
  public void setModel(VActor model) {
    this.model = model;
    init();
  }

  
  public VActor getModel() {
    return model;
  }

  public JButton getButton() {
    return button;
  }

  /**
   * Returns the actor action.
   */
  public Action getAction() {
    return action;
  }

  // --------------------------------------------------------------------
  // PRIVATE METHODS
  // --------------------------------------------------------------------

  /**
   * Loads the actor icon
   */
  private final ImageIcon loadImage(String iconName) {
    ImageIcon   image;

    image = Utils.getImage(iconName + ".png");

    if (image == null || image == Utils.UKN_IMAGE) {
      image = Utils.getImage(iconName + ".gif");
    }

    return image;
  }

  private void init() {
    action = new DActorAction(model.getMenuItem(),
                              (model.getIconName() != null) ?
                              loadImage(model.getIconName()) :
                              null);
    if (model.getAcceleratorKey() != KeyEvent.VK_UNDEFINED) {
      action.putValue(Action.ACCELERATOR_KEY,
                      KeyStroke.getKeyStroke(model.getAcceleratorKey(),
                                             model.getAcceleratorModifier()));
    }

    action.putValue(Action.SHORT_DESCRIPTION, model.getHelp());
    action.setEnabled(false);

    if (button == null) {
      button = new JMenuButton(action);
    }
  }

  // --------------------------------------------------------------------
  // DACTOR ACTION
  // --------------------------------------------------------------------

  private class DActorAction extends AbstractAction {

    // -----------------------------------------------
    // CONSTRUCTOR
    // -----------------------------------------------

    /*package*/ DActorAction(String name, Icon icon) {
      super(name, icon);
    }

    // ------------------------------------------------
    // ABSTRACT ACTION IMPLEMNTATION
    // ------------------------------------------------

    public void actionPerformed(ActionEvent e) {
      model.performAction();
    }

    // -------------------------------------------------
    // DATA MEMBERS
    // -------------------------------------------------

    /**
     * Comment for <code>serialVersionUID</code>
     */
    private static final long	serialVersionUID = -6510825866215273279L;
  }

  // ----------------------------------------------------------------------
  // DATA MEMBERS
  // ----------------------------------------------------------------------

  private JButton					button;
  private Action					action;
  private VActor					model;
}

File: galite-core/src/main/kotlin/org/kopi/galite/ui/visual/DActor.kt
/*
 * Copyright (c) 1990-2016 kopiRight Managed Solutions GmbH
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License version 2.1 as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * $Id: DActor.java 35180 2017-07-14 16:31:39Z hacheni $
 */
package org.kopi.galite.ui.visual

import com.vaadin.flow.component.html.Div
import com.vaadin.flow.component.html.Span
import org.kopi.galite.visual.UActor
import org.kopi.galite.visual.VActor
import java.awt.event.KeyEvent

/**
 * The `DActor` is the vaadin implementation of
 * the [UActor]. The actor can be represented by a [Button]
 * if it has a valid icon name.
 *
 *
 *
 * The actor action is handled by a [ShortcutListener] registered
 * of the [DWindow] which is the receiver of all actors actions.
 *
 */
class DActor(private var model: VActor) : Span(), UActor {
  // --------------------------------------------------
  // ACCESSORS
  // --------------------------------------------------
  override fun setModel(model: VActor) {
    this.model = model
  }

  override fun getModel(): VActor {
    return model
  }

  override fun isEnabled(): Boolean {
    return super.isEnabled()
  }

  override fun setEnabled(enabled: Boolean) {
    super.setEnabled(enabled)
  }

  override fun isVisible(): Boolean {
    return super.isVisible()
  }

  override fun setVisible(visible: Boolean) {
    super.setVisible(visible)
  }

  companion object {
    // --------------------------------------------------
    // PRIVATE METHODS
    // --------------------------------------------------
    /**
     * Creates the actor description.
     * @param model The actor model.
     * @return The actor description.
     */
    private fun getDescription(model: VActor): String? {
      return if (model.acceleratorKey > 0) {
        if (model.acceleratorModifier == 0) {
          model.help + " [" + KeyEvent.getKeyText(model.acceleratorKey) + "]"
        } else {
          model.help + " [" + KeyEvent.getKeyModifiersText(model.acceleratorModifier) + "-" + KeyEvent.getKeyText(model.acceleratorKey) + "]"
        }
      } else {
        model.help
      }
    }

    /**
     * Returns the corrected accelerator key.
     * @param acceleratorKey The original accelerator key.
     * @return The corrected accelerator key.
     */
    private fun correctAcceleratorKey(acceleratorKey: Int): Int {
      return if (acceleratorKey == 10) 13 else acceleratorKey
    }

    /**
     * Returns the corrected modifier accelerator key.
     * @param acceleratorModifier The original modifier accelerator key.
     * @return The corrected modifier accelerator key.
     */
    private fun correctAcceleratorModifier(acceleratorModifier: Int): Int {
      TODO()
    }
  }
  // --------------------------------------------------
  // CONSTRUCTOR
  // --------------------------------------------------
  /**
   * Creates the visible actor component from a given model.
   * @param model The **not null** model.
   */
  init {
    isEnabled = false

    val label = Span(model.actorIdent)
    add(label)


    model.setDisplay(this)
  }
}
